easynav_gps_vff/easynav_gps_vff_controller/src/easynav_gps_vff_controller/GpsVffController.cpp

```diff
@@ -0,0 +1,306 @@
+  marker_array_pub_ = node->create_publisher<visualization_msgs::msg::MarkerArray>(
+    "/markers_vff", 10);
+
+  return {};
+}
+
+void GpsVffController::imu_callback(const sensor_msgs::msg::Imu::SharedPtr msg)
+{
+  // Extract the yaw angle from the IMU data
+  tf2::Quaternion q(
+    msg->orientation.x, msg->orientation.y,
+    msg->orientation.z, msg->orientation.w);
+  tf2::Matrix3x3 m(q);
+  double roll, pitch, yaw;
+  m.getRPY(roll, pitch, yaw);
+  yaw_ = yaw;
+}
+
+geometry_msgs::msg::TwistStamped GpsVffController::get_cmd_vel()
+{
+  return cmd_vel_;
+}
+
+visualization_msgs::msg::MarkerArray
+GpsVffController::get_debug_vff(const VFFVectors & vff_vectors, std::string frame_id)
+{
+  visualization_msgs::msg::MarkerArray marker_array;
+
+  marker_array.markers.push_back(make_marker(vff_vectors.attractive, BLUE, frame_id));
+  marker_array.markers.push_back(make_marker(vff_vectors.repulsive, RED, frame_id));
+  marker_array.markers.push_back(make_marker(vff_vectors.result, GREEN, frame_id));
+
+  return marker_array;
+}
+
+visualization_msgs::msg::Marker
+GpsVffController::make_marker(
+  const std::vector<double> & vector, VFFColor vff_color,
+  std::string frame_id)
+{
+  visualization_msgs::msg::Marker marker;
+
+  marker.header.frame_id = frame_id;
+  marker.header.stamp = get_node()->now();
+  marker.type = visualization_msgs::msg::Marker::ARROW;
+  marker.id = visualization_msgs::msg::Marker::ADD;
+
+  geometry_msgs::msg::Point start;
+  start.x = 0.0;
+  start.y = 0.0;
+  geometry_msgs::msg::Point end;
+  start.x = vector[0];
+  start.y = vector[1];
+  marker.points = {end, start};
+
+  marker.scale.x = 0.05;
+  marker.scale.y = 0.1;
+
+  switch (vff_color) {
+    case RED:
+      marker.id = 0;
+      marker.color.r = 1.0;
+      break;
+    case GREEN:
+      marker.id = 1;
+      marker.color.g = 1.0;
+      break;
+    case BLUE:
+      marker.id = 2;
+      marker.color.b = 1.0;
+      break;
+  }
+  marker.color.a = 1.0;
+
+  return marker;
+}
+
+double GpsVffController::normalize_angle(double angle)
+{
+  angle = fmod(angle + M_PI, 2.0 * M_PI); // range [0, 2π)
+  if (angle < 0) {
+    angle += 2.0 * M_PI; // ensure positive
+  }
+  return angle - M_PI;
+}
+
+VFFVectors GpsVffController::get_vff(
+  double angle_error,
+  const pcl::PointCloud<pcl::PointXYZ> & pointcloud_,
+  std::string frame_id)
+{
+  // Init vectors
+  VFFVectors vff_vector;
+  // Attractive vector points toward the goal
+  vff_vector.attractive = {std::cos(angle_error), std::sin(angle_error)};
+  vff_vector.repulsive = {0.0, 0.0};
+  vff_vector.result = {0.0, 0.0};
+
+  // Find the closest obstacle in the point cloud
+  pcl::PointXYZ closest_point;
+  double min_distance = std::numeric_limits<double>::max();
+
+  for (const auto & point : pointcloud_) {
+    const double distance = std::hypot(point.x, point.y);
+    if (distance < min_distance) {
+      min_distance = distance;
+      closest_point = point;
+    }
+  }
```

Comment:
We should consider all the points within a working radius in a future improved version.